packages/core/src/electron-browser/window/electron-window-service.ts

```diff
@@ -33,20 +43,35 @@ export class ElectronWindowService extends DefaultWindowService {
 
     registerUnloadListeners(): void {
         window.addEventListener('beforeunload', event => {
-            // Either we can unload, or the user confirms that he wants to quit
-            if (this.canUnload() || this.shouldUnload()) {
-                // We are unloading
+            if (this.isUnloading) {
+                // Unloading process ongoing, do nothing:
+                return this.preventUnload(event);
+            } else if (this.closeOnUnload || this.canUnload()) {
```

Comment:
There's danger in checking the flag first:
1. This `canUnload` stops checking at the first contribution whose `onWillStop` returns truthy.
1. The user says yes, exit anyway, so we set the flag.
1. We exit, but any remaining contributions that would have run code on `onWillStop` are prevented from doing so because we don't check `canStop` again.

It's good, because we may not want to call `onWillStop` multiple times, but in this case, some contributions could be having `onWillStop` skipped entirely.

Reply:
Like you said the goal is to not call `canUnload` multiple times. But I am sorry I don't understand what I should do to address your other concern?

Reply:
https://github.com/eclipse-theia/theia/blob/ae35f9c1a4bf0127f8b15262422da93ad5a13ac0/packages/core/src/browser/window/default-window-service.ts#L51-L64

One option would be to change the implementation here so that any truthy return is saved, but not immediately returned. That way, `onWillStop` gets called at least once for each contribution. I'm not sure whether it's important to run the `onWillStop` in each contribution, but it's possible that the method is being used to do some cleanup, and it would be desirable not to skip it. Perhaps it's better not to make such a guarantee and just discourage use as a place where cleanup can happen, since the lifecycle won't be restarted if the user chooses not to exit, so the cleanup might be premature in any case.

Reply:
So do I understand correctly that this issue is independent from my change? I can address it here, but this wasn't in the original scope right?

Reply:
That is true; the same issue exists on master, since we always cancel the `canUnload` on the first truthy return and never rerun any `onWillStop` calls. So for the purposes of this PR, what you've done is fine.

Reply:
I remember that `onWillStop` was specifically meant to prevent unloading, and not so much to notify contributions. For this you should use the `WindowService.onUnload` event.

Reply:
https://github.com/eclipse-theia/theia/pull/8863